torch/lib/c10d/reducer.hpp

```diff
@@ -225,8 +229,12 @@ class TORCH_API Reducer {
   // the buckets
   void sync_bucket_indices(std::vector<std::vector<size_t>>& bucket_indices);
 
-  using GradCallback =
-      torch::distributed::autograd::DistAutogradContext::GradCallback;
+  using GradCallback = std::function<bool(at::Tensor&)>;
+#ifndef _WIN32
```

Comment:
Why do we need a `static_assert` specifically for Windows? And as far as I understand both template arguments to `std::is_same<>` are concrete types. Overall what is the purpose of this assertion?

Reply:
We "need" a static assert for when we're _not_ on Windows. The reasoning is: we want this `GradCallback` here to be an alias for whatever dist autograd's `GradCallback` type is, however under Windows we don't have access to dist autograd. Hence we "resolve" this local alias here to its concrete type, but we double-check that it still matches the dist autograd type when dist autograd is available. (We do it through a `static_assert`, rather than through a comment, because `static_assert`s are self-enforcing and never go stale).